src/main/java/com/google/step/servlets/HomePageServlet.java

```diff
@@ -0,0 +1,208 @@
+    response.getWriter().println(JsonFormatter.getHomePageJson(homePageMap));
+  }
+
+  /** Creates a list of list of deal maps for the home page */
+  private List<List<Map<String, Object>>> getHomePageDealMaps(List<List<Deal>> homePageDeals) {
+    List<List<Map<String, Object>>> homePageDealsMapList = new ArrayList<>();
+    for (List<Deal> dealList : homePageDeals) {
+      List<Map<String, Object>> homePageSectionDealMaps = new ArrayList<>();
+      for (Deal deal : dealList) {
+        User user = userManager.readUser(deal.posterId);
+        Restaurant restaurant = restaurantManager.readRestaurant(deal.restaurantId);
+        List<Tag> tags = tagManager.readTags(dealTagManager.getTagIdsOfDeal(deal.id));
+        int votes = voteManager.getVotes(deal.id);
+        Map<String, Object> homePageDealMap =
+            JsonFormatter.getBriefHomePageDealMap(deal, user, restaurant, tags, votes);
+        homePageSectionDealMaps.add(homePageDealMap);
+      }
+      homePageDealsMapList.add(homePageSectionDealMaps);
+    }
+    return homePageDealsMapList;
+  }
+
+  /** Retrieves deals posted by tags followed by user */
+  private List<Deal> getDealsPublishedByTags(Set<Long> tagIds) {
+    List<Long> dealIdResults = new ArrayList<>();
+    for (Long id : tagIds) {
+      List<Long> dealIdsWithTag = dealTagManager.getDealIdsWithTag(id);
+      dealIdResults.addAll(dealIdsWithTag);
+    }
+    // Get rid of duplicate dealID (Deals with multiple tags)
+    List<Long> dealsWithoutDuplicates = new ArrayList<>(new HashSet<>(dealIdResults));
+    return dealManager.readDeals(dealsWithoutDuplicates);
+  }
+
+  /** Sorts deals based on new (Newest to oldest) */
+  private List<Deal> sortDealsBasedOnNew(List<Deal> deals) {
+    Collections.sort(
+        deals,
+        new Comparator<Deal>() {
+          @Override
+          public int compare(Deal deal1, Deal deal2) { // TODO fix parse
+            return LocalDateTime.parse(deal2.creationTimeStamp)
+                .compareTo(LocalDateTime.parse(deal1.creationTimeStamp)); // Descending
+          }
+        });
+    return deals;
+  }
+
+  private double epochSeconds(String timestamp) {
+    LocalDateTime time = LocalDateTime.parse(timestamp);
+    long epoch = time.atZone(ZoneId.of(LOCATION)).toEpochSecond();
+    return epoch;
+  }
+
+  /**
+   * Calculates a hot score for each deal entity, which takes into account both the time and the
+   * amount of votes it got
+   */
+  private double calculateHotScore(Deal deal, int votes) {
+    double order = Math.log(Math.max(Math.abs(votes), 1));
+    int sign = 0;
+    if (votes > 0) sign = 1;
+    else if (votes < 0) sign = -1;
+    double seconds = epochSeconds((String) deal.creationTimeStamp) - OLDEST_DEAL_TIMESTAMP;
+    return sign * order + seconds / 45000;
+  }
+
+  /** Sorts deals based on value (hot score or votes) */
+  private List<Deal> sortDealsBasedOnValue(List<Deal> deals, String attribute) {
```

Comment:
you changed `getSortedDealsBasedOnValue()` to `sortDealsBasedOnHotScore()` in commit `4926868`, but somehow it got reverted in `f1c1478`. Please fix.

Reply:
yeah cause it reduces the number of getVotes calls, so i changed it back

Reply:
then i wont have to call get votes for calculating hot score again

Reply:
nvm i separated it back